### PR TITLE
feat: requestSpendPermission + getHash

### PR DESCRIPTION
**Pull request by fan-zhang-sv**
### _Summary_

# Add Spend Permission Public Utilities

## Overview
This PR introduces a complete set of public utilities for handling spend permissions in the account SDK. These utilities enable developers to create, sign, and manage spend permissions with proper EIP-712 typed data signing and on-chain hash generation.

#### 1. `requestSpendPermission()`
- **Purpose**: Requests user approval to create a new spend permission via EIP-712 signature
- **Features**: 
  - Opens wallet popup for user signature
  - Supports configurable allowance, period, start/end times
  - Auto-generates salt and handles defaults (100-year expiry, current start time)
  - Returns complete `SpendPermission` object with signature and metadata
- **Parameters**: account, spender, token, chainId, allowance, periodInDays, plus optional start/end/salt/extraData

#### 2. `getHash()`
- **Purpose**: Computes the unique on-chain hash for a spend permission
- **Features**:
  - Calls `SpendPermissionManager.getHash()` contract method
  - Required for permission identification, approval, and revocation
  - Handles BigInt conversions for allowance and salt
- **Parameters**: permission message object and chainId

#### 3. `createSpendPermissionTypedData()`
- **Purpose**: Creates EIP-712 compliant typed data structure
- **Features**:
  - Proper field ordering for contract compatibility
  - Automatic address checksumming via viem
  - Smart defaults (current time, 100-year expiry, random salt)
  - Converts periods from days to seconds

### Utility Functions
- `toTimestampInSeconds()`: Converts Date objects to Unix timestamps
- `add100Years()`: Calculates far-future expiry dates
- `getRandomHexString()`: Generates cryptographically secure random salts

### EIP-712 Compliance
- Correct domain structure with SpendPermissionManager details
- Proper type definitions matching smart contract expectations
- Field order optimized for hash calculation compatibility

### Type Safety
- Full TypeScript support with proper types
- Address validation and checksumming
- BigInt handling for large numbers

### Error Handling
- Client availability checks
- Provider error propagation
- Contract call failure handling

## Testing
- **100% test coverage** across all new functionality
- Comprehensive unit tests for:
  - EIP-712 typed data generation
  - User signature flow simulation
  - Contract hash retrieval
  - Error scenarios and edge cases
  - Parameter validation and type conversion

## Breaking Changes
None - this is purely additive functionality.

## Usage Example
```typescript
import { requestSpendPermission, getHash } from '@base-org/account/spend-permission';

// Request user approval for spend permission
const permission = await requestSpendPermission({
  account: '0x1234...',
  spender: '0x5678...',
  token: '0xUSDC...', 
  chainId: 8453,
  allowance: parseUnits('100', 6), // 100 USDC
  periodInDays: 30, // Monthly allowance
  provider
});

// Get the on-chain hash
const hash = await getHash({
  permission: permission.permission,
  chainId: 8453
});
```

## Files Changed
- ✅ Added `getHash.ts` + tests
- ✅ Added `requestSpendPermission.ts` + tests  
- ✅ Added `utils.ts` + tests
- ✅ Updated exports in `index.ts`

This implementation provides a complete, production-ready spend permission system with robust error handling, comprehensive testing, and developer-friendly APIs.

### _How did you test your changes?_

https://github.com/user-attachments/assets/a265c660-17bf-4e2d-b8d5-c127410a804e

### CODE REVIEW

File: packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ *   end: Date.now() + (86400 * 365),
+ *   salt: '0x1234567890abcdef...',
+ *   extraData: '0x'
+ * };
+ *
+ * const hash = await getHash(permission, 8453); // Base mainnet
+ * console.log('Permission hash:', hash);
+ * ```
+ */
+export const getHash = async (
```
Comment (stephancill) [inline packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts:43]: would be nice if we did this offline without calling the contract

File: packages/account-sdk/src/interface/public-utilities/spend-permission/utils.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    allowance: string;
+    period: number;
+    start: number;
+    end: number;
+    salt: string;
+    extraData: Hex;
+  };
+};
+
+export function generateSpendPermissionTypedData(
```
Comment (stephancill) [inline packages/account-sdk/src/interface/public-utilities/spend-permission/utils.ts:70]: wdyt of `createSpendPermissionTypedData`?

File: packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts (hunk 1 above)
Comment (fan-zhang-sv) [reply packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts:43]: hmm leaning towards calling contract tho
* still thinks SDK should be as lean as possible
* we can honor contract as the single source of truth
* this call happens the same time as user approving spend permission and in the background, so no performance issue

File: packages/account-sdk/src/interface/public-utilities/spend-permission/utils.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    primaryType: 'SpendPermission',
+    message: {
+      account: getAddress(account),
+      spender: getAddress(spender),
+      token: getAddress(token),
+      allowance: allowance.toString(),
+      period: 86400 * periodInDays,
+      start: toTimestampInSeconds(start ?? new Date()),
+      end: toTimestampInSeconds(end ?? add100Years(new Date())),
+      salt: salt ?? getRandomHexString(32),
```
Comment (fan-zhang-sv) [inline packages/account-sdk/src/interface/public-utilities/spend-permission/utils.ts:91]: @spencerstock wanna have you double check this logic, salt will always be 32 byte here

File: packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts (hunk 1 above)
Comment (stephancill) [reply packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts]: alright

btw are we sure this doesn't cause the popup to be blocked?

Comment (stephancill) [reply packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts]: am i understanding correctly that the SDK has to be in the connected state to get the hash?

Comment (fan-zhang-sv) [reply packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts]: > are we sure this doesn't cause the popup to be blocked?

ahhh i haven't tested yet, but 
* we use `promise.all` to fire both at the same time, i think we should be good?
* also with the new popup blocked dialog, the UX for it wouldn't be too bad i think

Comment (fan-zhang-sv) [reply packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts]: >  SDK has to be in the connected state to get the hash?

ya this is true, cause we only get the RPC url from wallet connection. But i think if dapp doesn't connect beforehand, they won't be having the spend permission hash to begin with

File: packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import { readContract } from 'viem/actions';
+import { SpendPermissionTypedData } from '../utils.js';
+
+/**
+ * Gets the hash of a spend permission from the SpendPermissionManager contract.
+ *
+ * This function calls the getHash method on the SpendPermissionManager contract
+ * to compute the unique hash for a given spend permission. This hash can be used
+ * to identify the permission on-chain and is required for operations like approval,
+ * revocation, and spending.
```
Comment (spencerstock) [inline packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts:15]: nit - I don't think the hash is required for these operations

File: packages/account-sdk/src/interface/public-utilities/spend-permission/methods/requestSpendPermission.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  start?: Date; // default to now
+  end?: Date; // default to never (start + 100 years)
+  salt?: string; // default to a random value by crypto.getRandomValues
+  extraData?: string; // default to '0x'
+};
+
+/**
+ * Requests user approval to create a new spend permission.
+ *
+ * This helper method opens a popup to ask the user to sign a spend permission,
```
Comment (spencerstock) [inline packages/account-sdk/src/interface/public-utilities/spend-permission/methods/requestSpendPermission.ts:23]:
```suggestion
 * This helper method opens a prompt to ask the user to sign a spend permission,

File: packages/account-sdk/src/interface/public-utilities/spend-permission/methods/requestSpendPermission.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+ * which allows a specified spender to spend tokens from the user's account up to
+ * a certain allowance within defined time periods.
```
Comment (spencerstock) [inline packages/account-sdk/src/interface/public-utilities/spend-permission/methods/requestSpendPermission.ts:25]: can we link out to docs from here? There's a lot of info to swallow in this docstring

File: packages/account-sdk/src/interface/public-utilities/spend-permission/utils.ts (hunk 3 above)
Comment (spencerstock) [reply packages/account-sdk/src/interface/public-utilities/spend-permission/utils.ts:91]: Looks good to me!

File: packages/account-sdk/src/interface/public-utilities/spend-permission/utils.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+ * Returns a new Date object that is 100 years after the provided date.
+ *
+ * @param baseDate - The starting Date.
+ * @returns A Date 100 years in the future.
+ */
+function add100Years(baseDate: Date): Date {
+  const result = new Date(baseDate);
+  result.setFullYear(result.getFullYear() + 100);
+  return result;
+}
```
Comment (spencerstock) [inline packages/account-sdk/src/interface/public-utilities/spend-permission/utils.ts:128]: prefer we just use max value: `281474976710655` or use 101 years. If we use 100 years we'll probably see a date 100 years from now since the FE will only show "never" if we're explicitly longer than 100 years at the time the FE computes the time window - which will be a number of seconds after this is computing the 100 year time window.

**Review by spencerstock — COMMENTED** [review]
Looks really great!

File: packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts (hunk 1 above)
Comment (stephancill) [reply packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts]: not ideal if someone wants to calculate the spend permission hash server side

Comment (stephancill) [reply packages/account-sdk/src/interface/public-utilities/spend-permission/methods/getHash.ts]: but the hash has limited use anyway so nbd

File: packages/account-sdk/src/interface/public-utilities/spend-permission/utils.ts (hunk 7 above)
Comment (fan-zhang-sv) [reply packages/account-sdk/src/interface/public-utilities/spend-permission/utils.ts]: makes sense updateddd

**Review by spencerstock — APPROVED** [review]
Very nice!

**Review by stephancill — APPROVED** [review]
